### PR TITLE
Make a few tweaks and fix some typos

### PR DESCRIPTION
**Pull request by joneschrisg**
r? @rocallahan

### CODE REVIEW

File: rr.latex [hunk 1]
```diff
@@ -67,7 +66,7 @@ With the basic functionality in place, we discovered that the main performance b
 our applications was the
 context switching required by using {\tt ptrace} to monitor system calls. We implemented
 a novel {\it system-call buffering} optimization to eliminate those context switches, dramatically
-reducing recording and replay overhead. This optimization relies on modern Linux kernel
+reducing recording and replay overhead on important real-world workloads. This optimization relies on modern Linux kernel
```
Comment (joneschrisg) [inline rr.latex:69]: "dramatically" on its own feels a little subjective and maybe hard to defend, but I think the results in the table are pretty dramatic so sticking with that :).  Also re-emphasizing that we're not talking about toy programs.

File: rr.latex [hunk 2]
```diff
@@ -112,7 +111,7 @@ operations, and future related operations, are replayed. We do create one replay
 recorded thread (not strictly necessary), and we create one replay address
 space (i.e. process) per recorded address space, along with matching memory mappings.
 
-With this design, the interface between user-space and the kernel is our recording boundary.
+With this design, our recording boundary is the interface between user-space and the kernel.
```
Comment (joneschrisg) [inline rr.latex:114]: Nothing wrong here, the sentence just scanned more easily for me in this order.

File: rr.latex [hunk 3]
```diff
@@ -366,6 +366,7 @@ Our results address the following questions:
 \item What are the impacts of the system-call buffering and block cloning optimizations?
 \item What is the impact of not having to instrument code?
 \item How much space do \system{} traces consume?
+\item [TODO:] What is the memory overhead of \system{} recording and replay, across different kinds of workloads?
```
Comment (joneschrisg) [inline rr.latex:369]: v. https://github.com/rocallahan/rr-paper/issues/2

File: rr.latex [hunk 4]
```diff
@@ -378,7 +379,7 @@ We present a variety of workloads to illuminate \system{}'s strengths and weakne
 
 \emph{octane} runs the Google Octane benchmark under the Mozilla Spidermonkey Javascript engine (Mercurial revision 9bd900888753). This illustrates performance on CPU-intensive code in a complex language runtime.
 
-\emph{htmltest} runs the Mozilla Firefox HTML forms tests. The harness is excluded from recording (using {\tt mach mochitest -f plain --debugger \system{} dom/html/test/forms}). This is an example from real-world usage. About 30\% of user-space CPU time is in the harness.
+\emph{htmltest} runs the Mozilla Firefox HTML forms tests (Mercurial revision 9bd900888753). The harness is excluded from recording (using {\tt mach mochitest -f plain --debugger \system{} dom/html/test/forms}). This is an example from real-world usage. About 30\% of user-space CPU time is in the harness.
```
Comment (joneschrisg) [inline rr.latex:382]: I assume your SM shell and FF were built from the same revision?

File: rr.latex [hunk 5]
```diff
@@ -408,7 +409,7 @@ Each test was run six times, discarding the first result and reporting the geome
 
 ``Record no-syscallbuf'' and ``Replay no-syscallbuf'' report overhead with the system-call buffering optimization disabled (which also disables block cloning). ``Record no-cloning'' reports overhead with just block cloning disabled.
 
-``DynamoRio'' reports the overhead of running the tests under the DynamoRio \cite{Bruening2012} (version 6.1.0) ``null tool'', to estimate a lower bound for the overhead of using dynamic code instrumentation as an implementation technique. (DynamoRio is reported to be among the fastest dynamic code instrumentation engines.)
+``DynamoRio-null'' reports the overhead of running the tests under the DynamoRio \cite{Bruening2012} (version 6.1.0) ``null tool'', to estimate a lower bound for the overhead of using dynamic code instrumentation as an implementation technique. (DynamoRio is reported to be among the fastest dynamic code instrumentation engines.)
```
Comment (joneschrisg) [inline rr.latex:412]: The "-null" is just to remind skimmers that we're not comparing to an R&R tool build on DynamoRio, only the null tool.

File: rr.latex [hunk 6]
```diff
@@ -434,13 +435,14 @@ Figure \ref{optimizations-chart} shows the impact of system-call buffering and b
...
+Figure \ref{dynamorio-chart} compares \system{} recording overhead with the overhead of DynamoRio's ``null tool'', which runs all code through the DynamoRio instrumentation engine but does not modify the code beyond whatever is necessary to maintain supervised execution; this represents a minimal-overhead code instrumentation configuration. \emph{cp} executes very little user-space code and DynamoRio's overhead is low on that workload. On \emph{make} and \emph{sambatest} DynamoRio overhead is similar to \system{} recording, even though on \emph{make} DynamoRio can utilize multiple cores. On \emph{htmltest} DynamoRio's overhead is very high, perhaps because that test runs a lot of Javascript with dynamically generated and modified machine code. Implementing record-and-replay on top of dynamic instrumentation would incur significant additional overhead, so we would expect the resulting system to have significantly higher overhead than \system{}
+\footnote{We reported DynamoRio's crash on our ``octane'' workload to the developers at [REDACTED].}.
```
Comment (joneschrisg) [inline rr.latex:443]: (Did you or are you planning to do this?)

File: rr.latex [hunk 7]
```diff
@@ -587,7 +589,7 @@ Supporting recording and replay of parallel shared-memory applications with pote
...
+The current state of Linux on commodity x86 CPUs enables single-core user-space record and replay with low overhead, without pervasive code instrumentation --- but only just. This is fortuitous; we use software and hardware features for purposes they were not designed to serve. It is also a recent development; five years ago {\tt seccomp-bpf} and the Linux file cloning APIs did not exist, and commodity architectures with a deterministic hardware performance counter usable from user-space had only just appeared (Intel Westmere)\footnote{Performance counters have been usable for kernel-implemented replay \cite{Dunlap2002,Olszewski2009} for longer, because kernel code can observe and compensate for events such as interrupts and page faults}. By identifying the utility of these features for record-and-replay, we hope that they will be supported by an increasingly broad range of future systems. Further fundamental improvements will likely require hardware support for record-and-replay of parallel programs.
```
Comment (joneschrisg) [inline rr.latex:592]: It's hard to see what changed here --- just an `s/relay/replay/` typo fix.

**Comment by joneschrisg** [comment]
I wish I could offer more suggestions, but I'm afraid I'm too close to the work and this is pretty close to what I would have written :).  Looks great!

Only remaining comment is that I wish we could work something like "fastest", "lowest overhead", "dramatically" into the abstract, but I couldn't find a good way.

File: rr.latex (hunk 4 above)
Comment (rocallahan) [inline rr.latex:382]: Yep

File: rr.latex (hunk 6 above)
Comment (rocallahan) [inline rr.latex:443]: I had not, but now I have :-). https://github.com/DynamoRIO/dynamorio/issues/1930

File: rr.latex [hunk 8]
```diff
@@ -434,13 +435,14 @@ Figure \ref{optimizations-chart} shows the impact of system-call buffering and b
...
+Figure \ref{dynamorio-chart} compares \system{} recording overhead with the overhead of DynamoRio's ``null tool'', which runs all code through the DynamoRio instrumentation engine but does not modify the code beyond whatever is necessary to maintain supervised execution; this represents a minimal-overhead code instrumentation configuration. \emph{cp} executes very little user-space code and DynamoRio's overhead is low on that workload. On \emph{make} and \emph{sambatest} DynamoRio overhead is similar to \system{} recording, even though on \emph{make} DynamoRio can utilize multiple cores. On \emph{htmltest} DynamoRio's overhead is very high, perhaps because that test runs a lot of Javascript with dynamically generated and modified machine code. Implementing record-and-replay on top of dynamic instrumentation would incur significant additional overhead, so we would expect the resulting system to have significantly higher overhead than \system{}
+\footnote{We reported DynamoRio's crash on our ``octane'' workload to the developers at [REDACTED].}.
 
-\subsection{Space Usage} \label{space-results}
+\subsection{Disk Space Usage} \label{space-results}
```
Comment (rocallahan) [inline rr.latex:445]: Think I'll change this to Storage Usage since SSDs aren't really disks.

File: rr.latex [hunk 9]
```diff
@@ -483,7 +485,7 @@ The different workloads have highly varying space consumption rates, but several
 As discussed in Section \ref{counters}, \system{} uses hardware performance
 counters to measure application progress. We need a counter that
 increments very frequently so that the pair (counter, general purpose registers)
-uniquely identifies a program execution point in practice, and it must be ``deterministic'' as described earlier.
+uniquely identifies a program execution point in practice, and it must be ``deterministic'' as described earlier
```
Comment (rocallahan) [inline rr.latex]: This change seems like a typo

**Comment by rocallahan** [comment]
Thanks a ton!!!

**Pull request merged**